src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,520 @@
+		case UI_SCRIPT_ANY: {
+			char text[35];
+			SAFE_SPRINTF(text, "Contains %d nested scripts", ctx->complexScripts[ctx->level].remainingScripts);
+
+			ui_displayPaginatedText(
+			        HEADER,
+			        text,
+			        this_fn
+			);
+		}
+		break;
+		case UI_SCRIPT_N_OF_K: {
+			{
+				// max possible length 67: "Reuqired signatures: x. Contains x nested scripts"
+				// where x is 2^32-1
+				char text[68];
+				SAFE_SPRINTF(text, "Required signatures: %u. Contains %u nested scripts", ctx->scriptContent.requiredScripts, ctx->complexScripts[ctx->level].remainingScripts);
+
+				ui_displayPaginatedText(
+				        HEADER,
+				        text,
+				        this_fn
+				);
+			}
+			break;
+		}
+		case UI_SCRIPT_INVALID_BEFORE:
+		case UI_SCRIPT_INVALID_HEREAFTER:
+			ui_displayUint64Screen(
+			        HEADER,
+			        ctx->scriptContent.timelock,
+			        this_fn
+			);
+			break;
+		default:
+			THROW(ERR_INVALID_STATE);
+		}
+	}
+
+	UI_STEP(DISPLAY_UI_STEP_RESPOND) {
+		io_send_buf(SUCCESS, NULL, 0);
+		ui_displayBusy();
+	}
+
+	UI_STEP_END(DISPLAY_UI_STEP_INVALID);
+}
+#undef HEADER
+#undef ASSERT_UI_SCRIPT_TYPE_SANITY
+
+#define UI_DISPLAY_SCRIPT(UI_TYPE)\
+	ctx->ui_scriptType = UI_TYPE;\
+	ctx->ui_step = DISPLAY_UI_STEP_POSITION;\
+	deriveScriptHash_display_ui_runStep();
+
+// Start complex native script
+
+static void deriveNativeScriptHash_handleAll(read_view_t* view)
+{
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE_WITH_CTX("");
+
+	nativeScriptHashBuilder_startComplexScript_all(&ctx->hashBuilder, ctx->complexScripts[ctx->level].remainingScripts);
+
+	ctx->complexScripts[ctx->level].ui_type = UI_SCRIPT_ALL;
+	UI_DISPLAY_SCRIPT(UI_SCRIPT_ALL);
+}
+
+static void deriveNativeScriptHash_handleAny(read_view_t* view)
+{
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE_WITH_CTX("");
+
+	nativeScriptHashBuilder_startComplexScript_any(&ctx->hashBuilder, ctx->complexScripts[ctx->level].remainingScripts);
+
+	ctx->complexScripts[ctx->level].ui_type = UI_SCRIPT_ANY;
+	UI_DISPLAY_SCRIPT(UI_SCRIPT_ANY);
+}
+
+static void deriveNativeScriptHash_handleNofK(read_view_t* view)
+{
+	// parse data
+	ctx->scriptContent.requiredScripts = parse_u4be(view);
+
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE_WITH_CTX("required scripts = %u, ", ctx->scriptContent.requiredScripts);
+
+	// validate that the received requiredScripts count makes sense
+	VALIDATE(ctx->complexScripts[ctx->level].remainingScripts >= ctx->scriptContent.requiredScripts, ERR_INVALID_DATA);
+
+	nativeScriptHashBuilder_startComplexScript_n_of_k(&ctx->hashBuilder, ctx->scriptContent.requiredScripts, ctx->complexScripts[ctx->level].remainingScripts);
+
+	ctx->complexScripts[ctx->level].ui_type = UI_SCRIPT_N_OF_K;
+	UI_DISPLAY_SCRIPT(UI_SCRIPT_N_OF_K);
+}
+
+static void deriveNativeScriptHash_handleComplexScriptStart(read_view_t* view)
+{
+	TRACE_BUFFER(view->begin, (size_t) (view->end - view->begin));
```

Comment:
There is a macro for this purpose: `VIEW_REMAINING_TO_TUPLE_BUF_SIZE`.

(Should be applied several times in the code.)

Reply:
I tried to do it this way when I was first implementing it, but for some reason it didn't compile.

Reply:
```suggestion
	TRACE_BUFFER(view->begin, VIEW_REMAINING_TO_TUPLE_BUF_SIZE(view));
```
This compiles for me.

Reply:
But this is not correct if I am not mistaken.
```c
// Note(ppershing): these macros expand to two arguments!
#define VIEW_REMAINING_TO_TUPLE_BUF_SIZE(view) (view)->ptr, view_remainingSize(view)
#define VIEW_PROCESSED_TO_TUPLE_BUF_SIZE(view) (view)->begin, view_processedSize(view)
```
the `TRACE_BUFFER` macro expects two arguments, the pointer to the buffer and the size. So I think the correct usage should be:
```c
TRACE_BUFFER(VIEW_REMAINING_TO_TUPLE_BUF_SIZE(view));
```
However, that doesn't compile for me. Because this code is repeated I suggest we could create a macro `TRACE_REMAINING_VIEW` or similar.

Or I can move the tracing to the `deriveNativeScriptHash_handleAPDU` method, but I didn't do it at first, because of the complex native script end call, which is now deprecated. So I will probably move the tracing to this top level function and then we don't have to repeat it.

Reply:
Interesting. It does not compile because the outer macro unrolls first, and it has just a single argument at the time. Well, macros were never all-powerful :)

I prefer the current approach with tracing in `deriveNativeScriptHash_handleAPDU`.